src/inset_state/flatten_ellipse_density.cpp

```diff
@@ -0,0 +1,488 @@
+    velocity[key] = Point(
+      flux_mp.at(key).x() / rho_mp.at(key),
+      flux_mp.at(key).y() / rho_mp.at(key));
+  }
+}
+
+Point interpolate(
+  const Point p,
+  const Delaunay &dt,
+  const std::unordered_map<Point, Point> &mp)
+{
+  // Find the triangle containing the point
+  const Face_handle fh = dt.locate(p);
+
+  // Get three vertices
+  const Point v1 = fh->vertex(0)->point();
+  const Point v2 = fh->vertex(1)->point();
+  const Point v3 = fh->vertex(2)->point();
+
+  // Calculate barycentric coordinates
+  std::tuple<Scd::FT, Scd::FT, Scd::FT> bary_coor;
+  bary_coor =
+    CGAL::Barycentric_coordinates::triangle_coordinates_in_tuple_2<Point>(
+      v1,
+      v2,
+      v3,
+      p);
+
+  // Get the barycentric coordinates
+  const double bary_x = std::get<0>(bary_coor);
+  const double bary_y = std::get<1>(bary_coor);
+  const double bary_z = std::get<2>(bary_coor);
+
+  // Get projected vertices
+  const Point v1_proj = mp.at(v1);
+  const Point v2_proj = mp.at(v2);
+  const Point v3_proj = mp.at(v3);
+
+  // Calculate projected point of p
+  const Point p_proj = Point(
+    bary_x * v1_proj.x() + bary_y * v2_proj.x() + bary_z * v3_proj.x(),
+    bary_x * v1_proj.y() + bary_y * v2_proj.y() + bary_z * v3_proj.y());
+  return p_proj;
+}
+
+void InsetState::flatten_ellipse_density()
+{
+  std::cerr << "In flatten_ellipse_density()" << std::endl;
+
+  // Get GeoDiv name from polygon id
+  std::vector<std::string> pgn_id_to_geo_id;
+
+  std::unordered_map<Point, double> rho_mp;
+  std::unordered_map<Point, Point> flux_mp;
+
+  // eul[i][j] will be the new position of proj_[i][j] proposed by a simple
+  // Euler step: move a full time interval delta_t with the velocity at time t
+  // and position (proj_[i][j].x, proj_[i][j].y)
+  std::unordered_map<Point, Point> eul;
+
+  // mid[i][j] will be the new displacement proposed by the midpoint
+  // method (see comment below for the formula)
+  std::unordered_map<Point, Point> mid;
+
+  // (vx_intp, vy_intp) will be the velocity at position (proj_.x, proj_.y) at
+  // time t
+  std::unordered_map<Point, Point> v_intp;
+
+  // (vx_intp_half, vy_intp_half) will be the velocity at the midpoint
+  // (proj_.x + 0.5*delta_t*vx_intp, proj_.y + 0.5*delta_t*vy_intp) at time
+  // t + 0.5*delta_t
+  std::unordered_map<Point, Point> v_intp_half;
+
+  std::unordered_map<Point, Point> velocity;
+
+  // Constants for the numerical integrator
+  const double inc_after_acc = 1.1;
+  const double dec_after_not_acc = 0.75;
+  const double abs_tol = (std::min(lx_, ly_) * 1e-6);
+
+  // Clear previous triangle transformation data
+  proj_qd_.triangle_transformation.clear();
+
+  for (const Point &pt : unique_quadtree_corners_) {
+    proj_qd_.triangle_transformation.insert_or_assign(pt, pt);
+  }
+
+  // TODO: Does the next line adjust rho_mean if the total inset area has
+  //       changed in previous runs of flatten_density() or
+  //       flatten_ellipse_density()?
```

Comment:
I believe the answer is effectively yes because the commented-out section in `main.cpp` indicates that we planned to project the coordinates after each run of `flatten_ellipse_density()`. We may want to repeat the comment in `fill_with_density.cpp`:

```
  // We assume that target areas that were zero or missing in the input have
  // already been replaced by
  // CartogramInfo::replace_missing_and_zero_target_areas().
  // We also correct for a drift in the total inset area, in case it is
  // present, by adjusting mean_density. The idea is to treat the exterior
  // area as if it were a polygon with target area
  // C * (lx * ly - initial_area_) and current area
  // C * (lx * ly - total_inset_area). The constant prefactor C is the same
  // in both areas and, thus, cancels out when taking the ratio.
```